### PR TITLE
Feature/survey email

### PR DESCRIPTION
**Pull request by XanderVertegaal**
This PR enables the user to add an email address when creating/editing a survey. When results from the survey are uploaded to the server, an email is sent out asynchronously to the provided address. The email looks like this.

image

Changes:
- A column `email_address` has been added to the `survey` table, and a migration has been written accordingly.
- The survey form is updated with an `email address` field, with validation and a description.
- The application and its database have been dockerized and combined in a Docker Compose network.
- Celery and Redis are added for async email handling.
- The README has been updated, documenting the new Docker approach.

NB: the SMTP server is not reachable from anywhere else but the server network. Therefore, sending emails will be difficult to test until we actually deploy to a test server.

### CODE REVIEW

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
After one hiccup, which I was able to solve quickly without help, `docker compose` ran smoothly and everything seems to work. Kudos on that!

This branch, PR and review combine two independent sets of changes: those pertaining to Dockerized deployment and those pertaining to emailing the survey results. I understand why you did it this way in this case, but generally, I would prefer that you submit two indepent PRs where neither contains changes related to the other. We can discuss the git-fu that makes this possible, if you like.

> NB: the SMTP server is not reachable from anywhere else but the server network. Therefore, sending emails will be difficult to test until we actually deploy to a test server.

The Flask-Mail documentation [discusses a way](https://pythonhosted.org/Flask-Mail/#unit-tests-and-suppressing-emails) in which emails can be intercepted in debug mode, so you can log them instead of sending them and/or check their contents in a unittest. Please explore this option; it would be reassuring to have a way to check that this feature works as intended, without having to deploy it.

The existing tests are based on [doctest](https://docs.python.org/3/library/doctest.html). You can see an example of an existing test in `coloringbook/utilities.py:subject_from_json`, lines 28-53. That particular test will currently not run, because you still need to add `coloringbook/utilities.py` to the `test.py`. Try `python test.py` once you have done that, then try writing your own test for `is_fill_correct` as a warmup exercise. With that and the Flask-Mail documentation, I think you should be able to write some test code for the email feature as well.

Apart from the comments above and below, the code changes look perfectly sensible to me. :+1:

File: coloringbook/admin/utilities.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# coding=utf-8
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply coloringbook/admin/utilities.py:1]: I'm afraid so. My container does not run without it, in any case.

image

(The character denoted with \xc3 is ã, apparently.)

**Comment by XanderVertegaal** [comment]
I'm glad to hear you got the docker containers up and running!

Your comments on splitting this branch and writing tests are taken to heart. I will close this PR, cherry pick the commits from this branch onto two new separate branches and open PRs for them. Your other comments will be addressed there as well.

File: Readme.md [hunk 2]
```diff
@@ -32,35 +32,46 @@ Researchers can compose their own surveys with custom images and sounds. The sur
...
+    MYSQL_PORT=1234
+    MYSQL_USER=abcdefg
+    MYSQL_PASSWORD=abcdefg
+    MYSQL_DB=abcdefg
+    MYSQL_ROOT_PASSWORD=abcdefg
 
-**Note** that you should not use the password quoted above. You can choose the name of the database and the name of the user freely as well.
-
-Coloring Book obtains the necessary information to connect to the database from a user-provided configuration file. This is what the file contents should minimally look like:
+Secondly, a file `CONFIG.py` is required with minimally the following contents
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply Readme.md:46]: The capitalization in CONFIG.py (in the previous PR) was based on my misreading of the README, which used all-caps "CONFIG" as a placeholder for 'the actual name of the config file'. There is no reason why the name of the configuration file should be in all caps. It has been fixed in the new PR.

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,5 @@
 *.pyc
-/config.py
+/coloringbook/config.py
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply .gitignore:2]: Lowercase is now used throughout (as it should be).

File: Readme.md [hunk 4]
```diff
@@ -32,35 +32,46 @@ Researchers can compose their own surveys with custom images and sounds. The sur
 
 On the client side, test subjects just need to run a HTML5 capable browser.
 
-On the server side, Coloring Book is a WSGI application, which means that it will run on any webserver that supports standard Python web applications. Apart from a WSGI-enabled webserver, install the dependencies using `pip install -r requirements.txt` or preferably `pip-sync` from the pip-tools package.
+Coloring Book is deployed using Docker Compose. The application needs two configuration files, which must be present in the same folder as `docker-compose.yml`.
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply Readme.md:35]: This was confusing indeed. I have corrected this in the new PR.

File: coloringbook/admin/utilities.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply coloringbook/admin/utilities.py:1]: This magic comment appears to be necessary. My container will not run without it.

image

(The character denoted with \xc3 is ã, apparently.)

File: migrations/alembic.ini [hunk 5]
```diff
@@ -2,11 +2,13 @@
...
 # template used to generate migration files
-# file_template = %%(rev)s_%%(slug)s
+file_template = %%(rev)s_%%(slug)s
 
 # set to 'true' to run the environment during
 # the 'revision' command, regardless of autogenerate
-# revision_environment = false
+; revision_environment = false
+
+script_location = %(here)s
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply migrations/alembic.ini:11]: I cannot create a migration without them. I have looked at this again, and the only required line is `script_location = %(here)s`; the other lines can be left as is. (I do not know what they do exactly.)

File: coloringbook/templates/email/email.html [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    <tbody>
+        {% for evaluation in result.evaluations %}
+        <tr>
+            <td>{{ evaluation.page }}</td>
+            <td>{{ evaluation.target }}</td>
+            <td>{{ evaluation.color }}</td>
+            <td>{{ evaluation.correct }}</td>
+        </tr>
+        {% else %}
+            <tr>Er zijn geen resultaten geregistreerd.</tr>
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply coloringbook/templates/email/email.html:29]: Toen ik het testte, zag het er goed uit, maar het kan inderdaad netter:

```
        {% else %}
            <tr>
                <td colspan="3">Er zijn geen resultaten geregistreerd.</td>
            </tr>
        {% endfor %}
```

File: coloringbook/templates/email/email.html [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        {% endfor %}
+    </tbody>
+</table>
+{% endfor %}
+
+<p>Hartelijke groeten,</p>
+
+<p>Het Coloring Bookteam</p>
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply coloringbook/templates/email/email.html:37]: Not really; good spot!

**Pull request closed without merging**